[perf] separate impls for refs to simplifiable types

lqd opened the pull request:
As discussed on zulip, some of the coherence checks on the `window` crates are hot on `RefSimplifiedType`s, I believe because the crate contains a lot of types with `From` conversions between them (emulating `COM` hierarchies IIUC), including references to these types.

This PR separates impls for references to simplifiable types from the current non-blanket impls, for faster lookups.

Opening as draft for feedback purposes, especially as to what kind of tests @lcnr would like to see in this PR.

lqd commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 89df198d30df99b9bc5c6738380f0b7b3759a1af with merge 8185d21406dfb1106547fc5cb973f6dffdc2b574...
<!-- homu: {"type":"TryBuildStarted","head_sha":"89df198d30df99b9bc5c6738380f0b7b3759a1af","merge_sha":"8185d21406dfb1106547fc5cb973f6dffdc2b574"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3782136892/jobs/6430846345)
Build commit: 8185d21406dfb1106547fc5cb973f6dffdc2b574 (`8185d21406dfb1106547fc5cb973f6dffdc2b574`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3782136892/jobs/6430846345"},"merge_sha":"8185d21406dfb1106547fc5cb973f6dffdc2b574"} -->

rust-timer commented:
Queued 8185d21406dfb1106547fc5cb973f6dffdc2b574 with parent 731e0bf721c1ec2c7843547e86b6157b40a437d1, future [comparison URL](https://perf.rust-lang.org/compare.html?start=731e0bf721c1ec2c7843547e86b6157b40a437d1&end=8185d21406dfb1106547fc5cb973f6dffdc2b574).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (8185d21406dfb1106547fc5cb973f6dffdc2b574): [comparison URL](https://perf.rust-lang.org/compare.html?start=731e0bf721c1ec2c7843547e86b6157b40a437d1&end=8185d21406dfb1106547fc5cb973f6dffdc2b574&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

@bors rollup=never
@rustbot label: -S-waiting-on-perf -perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=731e0bf721c1ec2c7843547e86b6157b40a437d1&end=8185d21406dfb1106547fc5cb973f6dffdc2b574&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-0.3%, -0.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.3% | [-0.3%, -0.3%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=731e0bf721c1ec2c7843547e86b6157b40a437d1&end=8185d21406dfb1106547fc5cb973f6dffdc2b574&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=731e0bf721c1ec2c7843547e86b6157b40a437d1&end=8185d21406dfb1106547fc5cb973f6dffdc2b574&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

lqd committed:
separate impls for refs to simplifiable types

Separate impls for references to simplifiable types to allow for faster lookups: the
set of possibly matching impls is smaller than if we stored them all as
`SimplifiedType::RefSimplifiedType`.

lqd committed:
clean-up `for_each_impl` to use `all_impls`

handle `impls_for_ref_x` and reduce duplication when iterating over all impls.

also: use it in `for_each_relevant_impl_treating_projections`

lqd committed:
return refs to simplifiable types in `non_blanket_impls_for_ty`

for users of the API, we return impls from `non_blanket_impls` and `impls_for_ref_x` as if they
weren't separated.

lqd commented:
I'm not sure this is complete just yet or if it's still missing some important changes, so I'll leave it as draft for now, and I'd also like to add some tests.

But since this topic showed up in the discussion to reduce hotspots for the atypical `windows-rs` project, I've tried a non-incremental `cargo check --all-features` for the [`windows` crate](https://github.com/microsoft/windows-rs/tree/master/crates/libs/windows): the try build `8185d21406dfb1106547fc5cb973f6dffdc2b574` was around 9% faster (330s vs 360s) compared to its parent `731e0bf721c1ec2c7843547e86b6157b40a437d1`.

As shown in the perf run above, these patterns are quite uncommon in our benchmark suite, and I believe in the wider ecosystem, but they do stress coherence, and could be present in other similar big crates bindings to existing type hierarchies (e.g. to other OS APIs, maybe UI toolkits with OOP flavors, etc).

With that in mind, r? @lcnr for when they come back from vacation.

lcnr commented on compiler/rustc_middle/src/ty/trait_def.rs:
```diff
@@ -147,8 +143,24 @@ impl<'tcx> TyCtxt<'tcx> {
+                if let Some(impls) = impls.impls_for_ref_x.get(&simplified_ref_ty) {
+                    return impls.iter().copied();
+                }
+            }
```
missing the `impls_for_ref_x` impls if you have `&T`. Doesn't matter with how this method is used but we should still be careful for it to be correct.

lcnr commented on compiler/rustc_middle/src/ty/trait_def.rs:
```diff
@@ -177,25 +189,65 @@ impl<'tcx> TyCtxt<'tcx> {
+                        return result;
+                    }
                 }
             }
```
the else branch is shared with `RefSimplifiedType`, isn't it? I think it makes sense to do
```rust 
if let Some(simp) = fast_reject::simplify_type(self, self_ty, TreatParams::AsPlaceholder) {
    impls.blanket_impls.iter()
     
    impls.non_blanket_impls.get(simp).iter()

    if let ty::Ref(_, ref_ty, mutablity) {
        if let Some(ref_simp) = fast_reject::simplify_type(self, *ref_ty, TreatParams::AsPlaceholder) {
            impls.impls_for_ref_x.get(ref_simp).iter()
        } else {
            impls.impls_for_ref_x.iter().flatten()
        }
    }
} else {
    self.all_impls(def_id)
}
```

lqd replied:
Good idea. I've followed this suggestion when rebasing.

lcnr reviewed (commented):
a few nits, after that

r=me

lcnr commented:
ah for tests:

probably:

`&_` correctly detecting impls for `&u32`.

coherence: `&T` overlapping with `&u32`

impls with `&u32` specializing an impl of `&T` and correctly using associated item definitions of the `&T` impl in the `&u32` impl.

That should cover all the edge cases I can think of rn

bors commented:
:umbrella: The latest upstream changes (presumably #108340) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd commented:
(rebased to discuss with lcnr, but the PR is still waiting on me)

@bors try

bors commented:
:hourglass: Trying commit 030f9bb8b266765e5b20354f54ee534e63ab7279 with merge cd2704ff205c1cca87105183ec3fc80182b8f5bb...
<!-- homu: {"type":"TryBuildStarted","head_sha":"030f9bb8b266765e5b20354f54ee534e63ab7279","merge_sha":"cd2704ff205c1cca87105183ec3fc80182b8f5bb"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/5137784963/jobs/9248183535)
Build commit: cd2704ff205c1cca87105183ec3fc80182b8f5bb (`cd2704ff205c1cca87105183ec3fc80182b8f5bb`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/5137784963/jobs/9248183535"},"merge_sha":"cd2704ff205c1cca87105183ec3fc80182b8f5bb"} -->

lqd commented:
Sanity check on `windows-rs` with the try build `cd2704ff205c1cca87105183ec3fc80182b8f5bb` and its parent `871b5952023139738f72eba235063575062bc2e9`, on `full` and `incr-full` scenarios:

```
Benchmark 1: CARGO_INCREMENTAL=0 cargo +winrs-try-parent check --all-features
  Time (mean ± σ):     336.616 s ±  0.582 s    [User: 327.807 s, System: 12.608 s]
  Range (min … max):   335.954 s … 337.044 s    3 runs

Benchmark 2: CARGO_INCREMENTAL=0 cargo +winrs-try check --all-features
  Time (mean ± σ):     309.976 s ±  1.063 s    [User: 301.125 s, System: 12.671 s]
  Range (min … max):   309.017 s … 311.119 s    3 runs

Summary
  'CARGO_INCREMENTAL=0 cargo +winrs-try check --all-features' ran
    1.09 ± 0.00 times faster than 'CARGO_INCREMENTAL=0 cargo +winrs-try-parent check --all-features'
```

```
Benchmark 1: cargo +winrs-try-parent check --all-features
  Time (mean ± σ):     533.431 s ±  0.708 s    [User: 511.599 s, System: 25.910 s]
  Range (min … max):   532.816 s … 534.205 s    3 runs

Benchmark 2: cargo +winrs-try check --all-features
  Time (mean ± σ):     477.190 s ±  0.913 s    [User: 456.594 s, System: 24.689 s]
  Range (min … max):   476.411 s … 478.195 s    3 runs

Summary
  'cargo +winrs-try check --all-features' ran
    1.12 ± 0.00 times faster than 'cargo +winrs-try-parent check --all-features'
```

lqd commented:
Coherence checks are on their way of being switched from the current solver to the next trait solver, via `-Ztrait-solver=next-coherence`. It performs the same in the rustc-perf benchmarks, and on `windows-rs`. 

However, this PR while being very beneficial to e.g. `windows-rs`, is very specific to the old solver architecture, and cannot be applied to the new solver. After discussing with @lcnr, we decided to close this PR because it will become obsolete as soon as the new solver is used for coherence, and landing prior to that, only for a short period, wouldn't be particularly useful.
